src/main.cpp

```diff
@@ -79,22 +85,55 @@ void parseArguments(int argc, char *argv[]);
 
 int main(int argc, char *argv[])
 {
+#if USE_MPI
+  int provided;
+  int localRank;
+
+  MPI_Init_thread(&argc, &argv, MPI_THREAD_FUNNELED, &provided);
+
+  if (provided < MPI_THREAD_FUNNELED)
+    MPI_Abort(MPI_COMM_WORLD, provided);
+
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+  MPI_Comm_size(MPI_COMM_WORLD, &procs);
+
+  // Each rank will run the benchmark on a single device
+  MPI_Comm shared_comm;
+  MPI_Comm_split_type(MPI_COMM_WORLD, MPI_COMM_TYPE_SHARED, 0,
```

Comment:
Why do we need to make a new communicator?

Reply:
because of `deviceIndex = localRank;` below.  this is the MPI way to figure out the MPI processes that share a node.  the ranks in this communicator can then be mapped to devices properly, assuming the size of `shared_comm` exceeds the number of devices per node.